### ISSUE TITLE
Page not found : Guac issue

### ISSUE THREAD

**Issue by CarsonHrusovsky** [kevoreilly/CAPEv2#1576]
## This is open source and you are getting __free__ support so be friendly!

# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [X] I am running the latest version
- [X] I did read the README!
- [X] I checked the documentation and found no answer
- [X] I checked to make sure that this issue has not already been filed
- [X] I'm reporting the issue to the correct repository (for multi-repository projects)
- [X] I have read and checked all configs (with all optional parts)

# Expected Behavior

View virtual machine over VNC using guac.

# Current Behavior

CAPE Sandbox redirects to Page not found upon trying to resolve 127.0.0.1:8000/guac/00/<sd>

## Steps to Reproduce

Please provide detailed steps for reproducing the issue.

1. step 1
Followed set-up procedure described in Read the docs for interactive session. Run installation script for guacamole. Within web.conf, activate guacamole. Within ESXi config, set up remoteconnection as detailed in this report for cuckoo-sandbox.
https://github.com/cuckoosandbox/cuckoo/issues/2319
Allowed connection to Windows 10 analysis machine over VNC.
2. step 2
confirm guacd and guac-web.service are running and functional.
3. you get it...
Upon attempting to access the interactive desktop mode, you are met with this:
![image](https://github.com/kevoreilly/CAPEv2/assets/95260807/963b5f86-122d-44ae-b136-608fa3f09db2)
(Essentially states: Using the URLconf defined in web.urls, Django tried these url patterns, in this order.)
No path is discovered.

Configuration:
[guacamole]
enabled = yes
mode = vnc
username =
password =
guacd_host = localhost
guacd_port = 4822
guest_protocol = vnc
guacd_recording_path = /opt/CAPEv2/storage/guacrecordings
guest_rdp_port = 3389

guacd.service 
June 02 19:20:14 userv systemd[1]: Started Guacamole Server.
June 02 19:20:14 userv guacd[232629]: Guacamole proxy daemon (guacd) version 1.5.0 started
June 02 19:20:14 userv guacd[232629]: guacd[232629]: INFO:        Guacamole proxy daemon (guacd) version 1.5.0 started
June 02 19:20:14 userv guacd[232629]: guacd[232629]: INFO:        Listening on host 127.0.0.1, port 4822
June 02 19:20:14 userv guacd[232629]: Listening on host 127.0.0.1, port 4822

guacd-web.service
No errors and application start up completes, running on 127.0.0.1:8008

## Context

I suspect my issues lay in the fact that I am using this through ESXi instead of KVM. I know this set up is possible with cuckoo and esxi using the setup I linked above. My next step will be to utilize KVM over ESXi as I believe this will be easiest solution, however if anyone has any advice on this setup for me to try I would be very thankful. 

| Question         | Answer
|------------------|--------------------
| Git commit       | Type `$ git log \| head -n1` to find out
| OS version       | Ubuntu 22.04, Windows 10, VMware ESXi

## Failure Logs

Please include any relevant log snippets or files here.
There are no logs generated by the guac service, and no failure logs are detected within CAPE when I run an analysis. No mention of any issues with CAPE Debug unrecognized_key are found. 

Virtual Machine is using VNCViewer and I have confirmed that it can be connected to via VNC. Is this a guacamole issue, django issue or VNC issue?

Any advice or help at all is greatly appreciated. Thank you!

**Comment by CarsonHrusovsky** [comment]
![image](https://github.com/kevoreilly/CAPEv2/assets/95260807/ad504dd3-b44f-4d1f-801a-9c9831a1883b)
For reference guac-web.service

**Comment by MarioRL0** [comment]
Are you using NGINX?
I had a similar issue weeks ago, maybe this can be useful to you https://github.com/kevoreilly/CAPEv2/issues/1508

**Comment by CarsonHrusovsky** [comment]
> Are you using NGINX? I had a similar issue weeks ago, maybe this can be useful to you #1508

Hello yes I saw your post and it did help me quite a bit, but I am still having the same issue. Maybe you can help me understand here, NGINX server is a proxy that wants to pass the localhost 127.0.0.1:8000 (CAPE Page) to gunicorn 127.0.0.1:8008 when you "browse" to the /guac page correct? My CAPE installation is running on localhost and I know gunicorn is running on localhost:8008, shouldn't this mean that if I manually browse to that interface, I should be greeted with either an error.html page or wait.html page? I had set up NGINX a few times and rewrote my config twice, but was still having the same issue, unsure if it could be NGINX but my config is essentially the same thing you see from the "Interactive Session" portion of the Readthedocs.

**Comment by MarioRL0** [comment]
Can you share the settings for your nginx config?

**Comment by CarsonHrusovsky** [comment]
![image](https://github.com/kevoreilly/CAPEv2/assets/95260807/d957fbc4-2ad5-42af-90f5-e798bad3bb66)
![image](https://github.com/kevoreilly/CAPEv2/assets/95260807/0dcaa14b-3ae1-4987-9fae-c4adaba5cfd1)

Here is my current config for nginx. I believe this seems to be functional.

**Comment by MarioRL0** [comment]
Are you accessing the web through your browser from 127.0.0.1:8000 or from 127.0.0.1:80?

It appears that you are currently accessing the web from 127.0.0.1:8000, whereas you should be accessing it from 127.0.0.1:80 as per your NGINX configuration.

![image](https://github.com/kevoreilly/CAPEv2/assets/91615276/b8a929f4-24af-4600-83ca-ffd20ca9b443)

**Comment by CarsonHrusovsky** [comment]
Thank you that was an issue I did not notice. However now it has thrown a few errors in my face, the most directly relevant is the following error message that is proposed after the fix you recommended above. Appreciate all the help on this! 
![image](https://github.com/kevoreilly/CAPEv2/assets/95260807/dd9f8397-bdcc-4d59-8f59-e8e0d5f4291d)

**Comment by MarioRL0** [comment]
Is there any extra info in cape-web or guac-web logs?

**Comment by CarsonHrusovsky** [comment]
![image](https://github.com/kevoreilly/CAPEv2/assets/95260807/14243d41-e863-4bc6-86f2-ea6601625c9b)
I assume this will take some digging but at least it is attempting to pull that vm (the rest of the cutoff text is just the name of the virtual machine). This is part of the configuration that is pretty murky as there is no documentation for this implementation on ReadtheDocs and the only thing am going off of is a forum post from 2018 for cuckoo.

**Comment by doomedraven** [comment]
you are welcome to extend that, we only suggest to use kvm and is uniq hypervisor that we support and focus on. Thanks @MarioRL0 for helping here

**Comment by CarsonHrusovsky** [comment]
@MarioRL0 any ideas on this issue here? Seems like maybe an issue of permissions but not completely sure. Had an issue with nginx too where my 127.0.0.1:80 page kept slowly losing CSS, started with favicons then completely lost css. Rewrote the config copying yours seems to be fixed (for now). If no ideas on my issue no worries I understand this is outside of the scope of what CAPE is officially developed for, appreciate all the help regardless :-)

**Comment by CarsonHrusovsky** [comment]
> I don't think I can help you with the error you're getting because in all the time I've been using CAPE, I haven't come across the same error. I'm sorry about that.

Hello again :) I have decided to completely scrap ESXi as I don't think my issue is getting resolved any time soon. I have set up KVM locally and everything seems okay except I am having the same error you encountered, blank screen instead of rendered VM through guacamole. Here is my nginx config (running on port 80 like you mentioned). No action is seen in guacamole console ever (I tried using the websocket-tunnel test but got nothing back). Finally here as well is the virtual machine configuration.
![image](https://github.com/kevoreilly/CAPEv2/assets/95260807/abf97896-20c3-4231-b33d-f8fbb5a7b3e2)
![image](https://github.com/kevoreilly/CAPEv2/assets/95260807/722fc333-0d9d-41cc-b082-160fa135f413)

**Comment by doomedraven** [comment]
in docs there is how to test this without nginx, so do that first so you can ensure that problems is not way much deeper

**Comment by CarsonHrusovsky** [comment]
> in docs there is how to test this without nginx, so do that first so you can ensure that problems is not way much deeper

![image](https://github.com/kevoreilly/CAPEv2/assets/95260807/d2826f21-2cd2-4822-8bdf-f1d5b8c72c3c)

I am met with this screen. Interestingly if I stop the nginx service and then attempt to make that connection via the same process but over port 8000 I am met with this. 

![image](https://github.com/kevoreilly/CAPEv2/assets/95260807/f2fcbbc4-e425-4a4a-ba3a-86c3d51238cc)


I've noticed with my django application that sometimes it doesn't have the correct permissions to display pages due to an icon on the page. For example, earlier it attempted to show me error.html when I was setting CAPE up with ESXi but the page couldn't load because it had an invalid block tag for "static". Upon just removing that line, which I will provide a screenshot of below, I was able to connect to the page without any CSS present. Do you think this may be related to my current issue? 

![image](https://github.com/kevoreilly/CAPEv2/assets/95260807/2e51b528-5af6-4789-aea3-f98265458f3f)

Upon removing all these elements from all those pages (error, wait and index), I am able to use a View Task button on the page, but that is it. Guacd.service shows no connection attempts at all. The most interesting thing that I will deep dive into is the guac-web.service status message upon attempting to open that connection (regardless of either being a real analysis or a test one using the test script you recommended). I will log into my github on my virtual machine and copy out the entire message as it doesn't fit on my screen currently, and will respond with that as well.

**Comment by CarsonHrusovsky** [comment]
Update my issue has been resolved! For anyone experiencing that same issue, here are the steps I took and how they resolved my issue. The most important part of this was that at least something loads on the guac page when you open interactive session. If anything except a blank screen opens you have additional issues. If the page opens blank, check the pages console like I did above, and see if you are getting that same error. If you aren't able to load the page and receive any errors in reference to "static" remove all CSS from all of the html pages found in the directory /opt/CAPEv2/web/guac/templates/guac. 
(Just comment out anything with the <link header). This helped me to fix my issue, additionally triple check your [ guacamole ] config found in the web.conf folder. I made changes to this in my testing that ended up causing me an additional day of headache. 

If you are reading this forum in hopes of getting ESXi to work, bad news for you I still have no idea what is wrong with that, its like guacamole cannot find that virtual machine despite having access for the analysis and all connection information is the same between them. Not sure if I will ever resolve that may take a look some time this week but for now KVM is the best move in my opinion!

**Comment by doomedraven** [comment]
Thanks for feedback, i guess you need to check the guac-main.js or
something like thay, as it was coded for kvm, so maybe the problem is there
with hour hypervisor

**Comment by CarsonHrusovsky** [comment]
> Thanks for feedback, i guess you need to check the guac-main.js or something like thay, as it was coded for kvm, so maybe the problem is there with hour hypervisor El mar, 13 jun 2023, 15:53, CarsonHrusovsky ***@***.***> escribió:

I was actually able to get guac working with ESXi by modifying a few of the python files found within /opt/CAPEv2/web/guac/. Namely consumers.py and views.py are needed to be edited to allow guac to make this connection. I can already make this connection in an ugly fashion by changing a few parts of the guac's GET request and running my KVM virtual machine locally at the same time as the actual machine used for analysis. I plan to put a post up here once I am done with a comprehensive guide on how to do this as I was unable to find any documentation before. 

I do have one question however, specifically within views.py, how does getting a DSN record go when making a connection to host? As a reference, if you go into /opt/CAPEv2/modules/machinery/esx.py you we see the following information in regards to making a connection. 

![image](https://github.com/kevoreilly/CAPEv2/assets/95260807/577b750f-4616-4710-ba6f-f651c20a9bbb)

I've tried to mirror this within guac's views.py but am not sure if this is exactly the correct format. Clearly this connection is already made once so it is possible. This connection is only needed by guac to tell if the virtual machine is on or off, and know what name to look for. I can make a VERY ugly hack to get this to work without needing this check, but it would be more smooth obviously if I could get guac working normally. Here is the code that I've modified (not working correctly yet) within /opt/CAPEv2/web/guac/views.py.

![image](https://github.com/kevoreilly/CAPEv2/assets/95260807/65697c79-7231-405b-937b-9efc9e7271e8)

I have ideas and will be testing, but if you have any additional insight I am all ears!
